main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,381 @@
+          case Right(a) => Right(a)
+          case Left((c, pats)) => Left((c, List(TypedAst.Pattern.Tag(c, c.name, TypedAst.Pattern.Tuple(pats, exp.tpe, exp.loc), exp.tpe, exp.loc))))
+        }
+      } else {
+        /* If the constructors are not complete, then we will fall to the wild/default case. In that case, we need to
+         * check for non matching patterns in the wild/default matrix.
+         */
+        FindNonMatchingPat(root, exp, defaultMatrix(rules), n - 1) match {
+          case Right(a) => Right(a)
+          case Left((c, pats)) => sigma match {
+            case Nil => Left(c, Pattern.Wild(exp.tpe, exp.loc) :: pats)
+            case _ => Left(c, List(TypedAst.Pattern.Tag(c, c.name, TypedAst.Pattern.Tuple(pats, exp.tpe, exp.loc), exp.tpe, exp.loc)))
+          }
+        }
+      }
+    }
+
+    /**
+      * Specialize a matrix of patterns for the Constructor ctor For a constructor of C(r1,...ra) and a matrix of width
+      * n, we return a matrix of width n+a-1
+      *
+      * e.g. If we have
+      * enum Option {
+      *    case Some a,
+      *    case Nothing
+      *  }
+      *
+      * And the pattern matrix
+      *
+      * case Some True =>
+      * case Some False =>
+      * case Nothing =>
+      *
+      * Specializing for Some gives use the matrix
+      * True =>
+      * False =>
+      *
+      * Where' we've taken a matrix had made of the arguments to the Some rows. Specializing for Nothing would give
+      * an empty matrix as it has no arguments.
+      *
+      * @param root The ast root
+      * @param ctor The constructor to specialize for
+      * @param rules The rules matrix to specialize
+      * @return The specialized matrix of rules
+      */
+    def Specialize(root: TypedAst.Root, ctor: TypedAst.Pattern, rules: List[List[TypedAst.Pattern]]): List[List[TypedAst.Pattern]] = {
+      // First figure out how many arguments are needed by the ctor
+      val numArgs = countCtorArgs(ctor)
+
+      def specializeRow(pat: List[TypedAst.Pattern], acc: List[List[TypedAst.Pattern]]) = pat.head match {
+        // A wild constructor is the same as the constructor
+        // with all its arguments as wild
+        case a: TypedAst.Pattern.Wild =>
+          (List.fill(numArgs)(a) ++ pat.tail) :: acc
+        case a: TypedAst.Pattern.Var =>
+          (List.fill(numArgs)(a) ++ pat.tail) :: acc
+
+        // If it's a pattern with the constructor that we are
+        // specializing for, we break it up into it's arguments
+        // If it's not our constructor, we ignore it
+        case TypedAst.Pattern.Tag(_, tag, exp, _, _) => exp match {
+          case TypedAst.Pattern.Tuple(elms, _, _) =>
+            if (tag.equals(patName(ctor))) {
+              (elms ++ pat.tail) :: acc
+            } else acc
+          case TypedAst.Pattern.Unit(_) =>
+            if (tag.equals (patName (ctor) ) ) {
+              pat.tail :: acc
+            } else acc
+          case _ =>
```

Comment:
If possible, we avoid wildcard patterns in the Flix compiler because that allows us to add new AST nodes and be sure that we get compiler warnings for unmatched cases :)